src/ngraph/op/shape_of.cpp

```diff
@@ -112,15 +137,22 @@ namespace
                 }
             }
 
-            replacements = OutputVector{std::make_shared<op::Concat>(dimensions, 0)};
+            replacement = std::make_shared<op::Concat>(dimensions, 0);
+            return true;
         }
-        return replacements;
+        return false;
     }
 }
 
-OutputVector op::v3::ShapeOf::constant_fold()
+bool op::v3::ShapeOf::evaluate(const EvaluatorTensorVector& output_values,
+                               const EvaluatorTensorVector& input_values)
+{
+    return evaluate_shape_of(output_values[0], input_values[0]);
+}
+
+bool op::v3::ShapeOf::constant_fold(OutputVector& output_values, const OutputVector& input_values)
 {
-    return constant_fold_shape_of(shared_from_this(), m_output_type, m_is_foldable);
+    return constant_fold_shape_of(this, output_values[0], input_values[0], m_is_foldable);
```

Comment:
More I think, the more I tend to agree with your design. But let's view it as follow: `evaluate` is a real reference kernel (for education purposes and validation), while `constant_fold` is for real-time execution in the application while graph transformation. Does it fit your intentions? Does it look fine with you?

Reply:
Note that I have not changed what the constant folder for ShapeOf does, just changed where it does it.

Reply:
One difference with constant-fold, exemplified by the one in ShapeOf, is that it might do partial evaluation, i.e. replace a node with a sub-graph that is simpler to evaluate at runtime. Evaluate has to come up with actual tensors.